### PR TITLE
Health factor tests

### PR DESCRIPTION
**Pull request by dmytro-shash**

### CODE REVIEW

File: contracts/controller/src/healthfactor.rs [hunk 1]
```diff
@@ -116,18 +121,24 @@ mod tests {
...
-            fraction_digits: 4,
-        });
-        prices.push(Price {
-            ticker_id: ticker_id_eth,
-            value: U128(20000),
-            volatility: U128(100),
-            fraction_digits: 4,
-        });
+
+        if near_price != 0 {
```
Comment (markblaize) [inline contracts/controller/src/healthfactor.rs:125]: @dmytro-shash you cannot change existing tests (until they are broken). If you need test without zero price you may create separate fixuture for that

File: contracts/controller/src/healthfactor.rs [hunk 2]
```diff
@@ -195,4 +209,160 @@ mod tests {
             "Health factor calculation has been failed"
         );
     }
+
+    #[test]
+    fn scenario_1() {
```
Comment (markblaize) [inline contracts/controller/src/healthfactor.rs:214]: @dmytro-shash please give a good names to those scenarios

File: contracts/controller/src/healthfactor.rs [hunk 3]
```diff
@@ -195,4 +209,160 @@ mod tests {
...
+        controller_contract.increase_borrows(
+            user_account.clone(),
+            AccountId::new_unchecked("dweth.near".to_string()),
+            WBalance::from(0),
+        );
+
+        assert_eq!(
+            controller_contract.get_health_factor(user_account.clone()),
+            controller_contract.get_health_threshold(),
+            "Test for account w/o collaterals and borrows has been failed"
```
Comment (markblaize) [inline contracts/controller/src/healthfactor.rs:233]: @dmytro-shash this error message seem like a good name for test

Comment (markblaize) [inline contracts/controller/src/healthfactor.rs:233]: @dmytro-shash this error message seem like a good name for test

**Review by markblaize — CHANGES_REQUESTED** [review]

File: contracts/controller/src/healthfactor.rs (hunk 1 above)
Comment (dmytro-shash) [reply contracts/controller/src/healthfactor.rs:125]: I meant to add the price of eth/near if they needed ( price are not  = 0 )

Comment (markblaize) [reply contracts/controller/src/healthfactor.rs:125]: there is few problems 

1) you may want to construct `prices` like this 
```
let prices: Vec<Price> = vec![ Price {..}, ...];
```
in this case you stick to RAII and immutability principles

2) it's not clear for the user of `init_price_volatility` why price where not added (someone may want add zero prices for some cases)

3) `init_price_volatility` has now actually 4 version of behavior (near_price == 0, near_price != 0, eth_price == 0, eth_price != 0) which in case of fixture is not that bad but in case of regular function requires 4x more tests

so if zero prices are not a problem or if you can implement separate function without prices and with it shuold be preferable

**Review by markblaize — CHANGES_REQUESTED** [review]

File: contracts/controller/src/healthfactor.rs (hunk 1 above)
Comment (dmytro-shash) [reply contracts/controller/src/healthfactor.rs]: solved 🙏🏻

File: contracts/controller/src/healthfactor.rs [hunk 4]
```diff
@@ -195,4 +254,160 @@ mod tests {
             "Health factor calculation has been failed"
         );
     }
+
+    #[test]
+    fn test_account_no_collaterals_and_borrows() {
```
Comment (pahomas) [inline contracts/controller/src/healthfactor.rs:259]: test_health_factor_wo_s_or_b

File: contracts/controller/src/healthfactor.rs [hunk 5]
```diff
@@ -195,4 +254,160 @@ mod tests {
...
+
+        assert_eq!(
+            controller_contract.get_health_factor(user_account.clone()),
+            controller_contract.get_health_threshold(),
+            "Test for account w/o collaterals and borrows has been failed"
+        );
+    }
+
+    #[test]
+    fn test_supply_100eth() {
```
Comment (pahomas) [inline contracts/controller/src/healthfactor.rs:283]: test_health_factor_s_100eth

File: contracts/controller/src/healthfactor.rs [hunk 6]
```diff
@@ -195,4 +254,160 @@ mod tests {
...
+            AccountId::new_unchecked("dweth.near".to_string()),
+            WBalance::from(0),
+        );
+
+        // Ratio that represents 150%
+        assert_eq!(controller_contract.get_health_factor(user_account), 15000);
+    }
+
+    #[test]
+    fn test_supply_100eth_borrow_70eth() {
```
Comment (pahomas) [inline contracts/controller/src/healthfactor.rs:304]: est_health_factor_with_supply_and_borrow

File: contracts/controller/src/healthfactor.rs [hunk 7]
```diff
@@ -195,4 +254,160 @@ mod tests {
...
+            AccountId::new_unchecked("dweth.near".to_string()),
+            WBalance::from(70),
+        );
+
+        // Ratio that represents 142.85%
+        assert_eq!(controller_contract.get_health_factor(user_account), 14285);
+    }
+
+    #[test]
+    fn test_supply_200eth_borrow_100near_for_price_1() {
```
Comment (pahomas) [inline contracts/controller/src/healthfactor.rs:325]: test_health_factor_with_supply_and_borrow_scenario_2

File: contracts/controller/src/healthfactor.rs [hunk 8]
```diff
@@ -195,4 +254,160 @@ mod tests {
...
+            AccountId::new_unchecked("dwnear.near".to_string()),
+            WBalance::from(100),
+        );
+
+        // Ratio that represents 200%
+        assert_eq!(controller_contract.get_health_factor(user_account), 20000);
+    }
+
+    #[test]
+    fn test_supply_200eth_borrow_100near_for_price_2() {
```
Comment (pahomas) [inline contracts/controller/src/healthfactor.rs:346]: test_health_factor_with_supply_and_borrow_scenario_3

File: contracts/controller/src/healthfactor.rs [hunk 9]
```diff
@@ -195,4 +254,160 @@ mod tests {
...
+            AccountId::new_unchecked("dwnear.near".to_string()),
+            WBalance::from(100),
+        );
+
+        // Ratio that represents 50%
+        assert_eq!(controller_contract.get_health_factor(user_account), 5000);
+    }
+
+    #[test]
+    fn test_supply_200near_borrow_100near() {
```
Comment (pahomas) [inline contracts/controller/src/healthfactor.rs:367]: test_health_factor_with_supply_and_borrow_scenario_4

File: contracts/controller/src/healthfactor.rs [hunk 10]
```diff
@@ -195,4 +254,160 @@ mod tests {
...
+            AccountId::new_unchecked("dwnear.near".to_string()),
+            WBalance::from(100),
+        );
+
+        // Ratio that represents 225%
+        assert_eq!(controller_contract.get_health_factor(user_account), 22500);
+    }
+
+    #[test]
+    fn test_supply_200eth_borrow_50eth_100near() {
```
Comment (pahomas) [inline contracts/controller/src/healthfactor.rs:388]: test_health_factor_with_supply_and_multi-borrow

**Review by pahomas — APPROVED** [review]

**Review by pahomas — APPROVED** [review]

File: contracts/controller/src/healthfactor.rs [hunk 11]
```diff
@@ -196,4 +255,160 @@ mod tests {
...
+            AccountId::new_unchecked("dwnear.near".to_string()),
+            WBalance::from(100),
+        );
+
+        // Ratio that represents 200%
+        assert_eq!(controller_contract.get_health_factor(user_account), 20000);
+    }
+
+    #[test]
+    fn test_health_factor_with_supply_and_borrow_scenario_3() {
```
Comment (markblaize) [inline contracts/controller/src/healthfactor.rs:347]: @dmytro-shash 

Scenario 5: 
  Given supply is 200 WETH
  And borrow is 100 WNEAR
  And Price is {ticker_id: WETH, value: 10000, volatility: 100%, fraction_digits: 4}
  And Price is {ticker_id: WNEAR, value: 10000, volatility: 100%, fraction_digits: 4}
  Then health_factor equal 200%
  When Price is {ticker_id: WETH, value: 5000, volatility: 100%, fraction_digits: 4}
  And Price is {ticker_id: WNEAR, value: 20000, volatility: 100%, fraction_digits: 4}
  Then health_factor equal 50%

File: contracts/controller/src/healthfactor.rs [hunk 12]
```diff
@@ -196,4 +255,160 @@ mod tests {
...
+            AccountId::new_unchecked("dwnear.near".to_string()),
+            WBalance::from(100),
+        );
+
+        // Ratio that represents 50%
+        assert_eq!(controller_contract.get_health_factor(user_account), 5000);
+    }
+
+    #[test]
+    fn test_health_factor_with_supply_and_borrow_scenario_4() {
```
Comment (markblaize) [inline contracts/controller/src/healthfactor.rs:368]: @dmytro-shash 

Scenario 6: 
  Given supply is 200 WETH
  And borrow is 100 WNEAR
  And Price is {ticker_id: WETH, value: 10000, volatility: 100%, fraction_digits: 4}
  And Price is {ticker_id: WNEAR, value: 10000, volatility: 100%, fraction_digits: 4}
  Then health_factor equal 200%
  When Price is {ticker_id: WETH, value: 10000, volatility: 90%, fraction_digits: 4}
  And Price is {ticker_id: WNEAR, value: 10000, volatility: 80%, fraction_digits: 4}
  Then health_factor equal 225%

**Review by markblaize — CHANGES_REQUESTED** [review]

**Review by pahomas — APPROVED** [review]

**Review by markblaize — APPROVED** [review]
